paddle/fluid/operators/collective/c_allreduce_op.h

```diff
@@ -115,36 +117,47 @@ class CAllReduceOpASCENDKernel : public framework::OpKernel<T> {
  public:
   void Compute(const framework::ExecutionContext& ctx) const override {
 #if defined(PADDLE_WITH_ASCEND_CL)
+    #define PRE_MALLOC_SIZE_BYTES 512
+
     auto in = ctx.Input<framework::LoDTensor>("X");
     auto out = ctx.Output<framework::LoDTensor>("Out");
-
     auto place = ctx.GetPlace();
     hcclDataType_t dtype = platform::ToHCCLDataType(in->type());
-
     int64_t numel = in->numel();
-    void* sendbuff = reinterpret_cast<void*>(const_cast<T*>(in->data<T>()));
-    // void* sendbuff = reinterpret_cast<void*>(const_cast<T*>(in->mutable_data<T>(place)));
 
-    out->Resize(in->dims());
-    // void* recvbuff = reinterpret_cast<void*>(const_cast<T*>(out->data<T>()));
-    void* recvbuff = reinterpret_cast<void*>(const_cast<T*>(out->mutable_data<T>(place)));
-    // void* recvbuff = sendbuff;
+    int64_t pre_tmp_size = PRE_MALLOC_SIZE_BYTES / sizeof(T);
+    int64_t tmp_numel = numel + pre_tmp_size * 2;
+
+    paddle::framework::LoDTensor tmp_in, tmp_out;
+    tmp_in.Resize({1, tmp_numel});
+    tmp_out.Resize({1, tmp_numel});
+    tmp_in.mutable_data<T>(place);  // allocate
+    tmp_out.mutable_data<T>(place);  // allocate
+
+    void* sendbuff = reinterpret_cast<void*>(const_cast<T*>(tmp_in.data<T>() + pre_tmp_size));
+    void* recvbuff = reinterpret_cast<void*>(const_cast<T*>(tmp_out.data<T>() + pre_tmp_size));
+
     std::string tag = ctx.Attr<std::string>("tag");
     int ring_id = ctx.Attr<int>("ring_id");
-    // s他的：
     std::string group = std::string(HCOM_GROUP_PREFIX) + std::to_string(ring_id);
-     group = "hccl_world_group";// std::string(HCOM_GROUP_PREFIX) + std::to_string(ring_id);
-
     auto comm = paddle::platform::HCCLCommContext::Instance().Get(ring_id, place);
 
     aclrtStream stream = nullptr;
+    auto dev_ctx = platform::DeviceContextPool::Instance().Get(place);
     if (ctx.Attr<bool>("use_calc_stream")) {
-      auto dev_ctx = platform::DeviceContextPool::Instance().Get(place);
       stream = static_cast<platform::NPUDeviceContext*>(dev_ctx)->stream();
     } else {
       stream = comm->stream();
     }
 
+    auto npu_place = BOOST_GET_CONST(platform::NPUPlace, place);
+
+    memory::Copy(npu_place, sendbuff, 
+                 npu_place, reinterpret_cast<void*>(const_cast<T*>(in->data<T>())),
+                 numel * sizeof(T),
+                 stream);
+    dev_ctx->Wait();
```

Comment:
Maybe this wait is not needed, you can test that.

Reply:
yeah, it run more smoothly after removed that.